bpfd/src/command.rs

```diff
@@ -231,176 +165,40 @@ impl TryFrom<AyaProgInfo> for KernelProgramInfo {
-}
-
-#[derive(Debug, Clone, Deserialize, Serialize)]
-pub(crate) struct TracepointAttachInfo {
-    pub(crate) tracepoint: String,
-}
-
-#[derive(Debug, Clone, Deserialize, Serialize)]
-pub(crate) struct KprobeAttachInfo {
-    pub(crate) fn_name: String,
-    pub(crate) offset: u64,
-    pub(crate) retprobe: bool,
-    pub(crate) namespace: Option<String>,
-}
-
-#[derive(Debug, Clone, Deserialize, Serialize)]
-pub(crate) struct UprobeAttachInfo {
-    pub(crate) fn_name: Option<String>,
-    pub(crate) offset: u64,
-    pub(crate) target: String,
-    pub(crate) retprobe: bool,
-    pub(crate) pid: Option<i32>,
-    pub(crate) namespace: Option<String>,
-}
-
-#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Serialize, Deserialize, Clone)]
-pub(crate) struct Metadata {
-    pub(crate) priority: i32,
-    pub(crate) attached: bool,
-}
-
-impl Metadata {
-    pub(crate) fn new(priority: i32) -> Self {
-        Metadata {
-            priority,
-            attached: false,
-        }
-    }
-}
-
-#[derive(Serialize, Deserialize, Clone)]
-pub(crate) enum Program {
-    Xdp(XdpProgram),
-    Tracepoint(TracepointProgram),
-    Tc(TcProgram),
-    Kprobe(KprobeProgram),
-    Uprobe(UprobeProgram),
-}
-
-#[derive(Serialize, Deserialize, Clone)]
-pub(crate) struct XdpProgram {
-    pub(crate) data: ProgramData,
-    pub(crate) info: XdpProgramInfo,
-}
-
-impl XdpProgram {
-    pub(crate) fn new(data: ProgramData, info: XdpProgramInfo) -> Self {
-        Self { data, info }
-    }
-}
-
-#[derive(Serialize, Deserialize, Clone)]
-pub(crate) struct TcProgram {
-    pub(crate) data: ProgramData,
-    pub(crate) info: TcProgramInfo,
-}
-
-#[derive(Serialize, Deserialize, Clone)]
-pub(crate) struct TracepointProgram {
-    pub(crate) data: ProgramData,
-    pub(crate) info: TracepointProgramInfo,
-}
-
-impl TracepointProgram {
-    pub(crate) fn new(data: ProgramData, info: TracepointProgramInfo) -> Self {
-        Self { data, info }
-    }
-}
-
-#[derive(Serialize, Deserialize, Clone)]
-pub(crate) struct KprobeProgram {
-    pub(crate) data: ProgramData,
-    pub(crate) info: KprobeProgramInfo,
-}
-
-#[derive(Serialize, Deserialize, Clone)]
-pub(crate) struct UprobeProgram {
-    pub(crate) data: ProgramData,
-    pub(crate) info: UprobeProgramInfo,
-}
-
-impl KprobeProgram {
-    pub(crate) fn _new(data: ProgramData, info: KprobeProgramInfo) -> Self {
-        Self { data, info }
-    }
-}
-
-impl UprobeProgram {
-    pub(crate) fn _new(data: ProgramData, info: UprobeProgramInfo) -> Self {
-        Self { data, info }
-    }
-}
-
-// ProgramData represents all of the core information needed to load
-// a program reguardless of ProgramType.
-#[derive(Serialize, Deserialize, Clone)]
+/// ProgramInfo stores information about bpf programs that are loaded and managed
+/// by bpfd.
+#[derive(Debug, Serialize, Deserialize, Clone)]
 pub(crate) struct ProgramData {
+    // known at load time, set by user
+    pub(crate) name: String,
```

Comment:
Make these fields private and add setters/getters for them. This will force us into better habits because we can return return borrows.

E.g `&str` instead of `String`

The argument for setters is:

```rust
program.data_mut().foo = 1;
```

vs.

```rust
program.data_mut().set_foo(1);
```

Reply:
Done! I love the better habbit lol so many less clones